source/particle/generator/probability_density_function.cc

```diff
@@ -0,0 +1,295 @@
+                               "function is positive in at least a part of the domain, also check the syntax of the function."));
+
+        // Determine the starting weight of this process, which is the sum of
+        // the weights of all processes with a lower rank
+        double start_weight = 0.0;
+        MPI_Scan(&local_function_integral, &start_weight, 1, MPI_DOUBLE, MPI_SUM, this->get_mpi_communicator());
+        start_weight -= local_function_integral;
+
+        // Build the map between integrated probability density and cell
+        std::map<double, types::LevelInd> cell_weights;
+        typename DoFHandler<dim>::active_cell_iterator cell = this->get_dof_handler().begin_active(),
+                                                       endc = this->get_dof_handler().end();
+        for (unsigned int i = 0; cell!=endc; ++cell)
+          if (cell->is_locally_owned())
+            {
+              // adjust the cell_weights according to the local starting weight
+              cell_weights.insert(std::make_pair(accumulated_cell_weights[i] + start_weight, types::LevelInd(cell->level(),cell->index())));
+              ++i;
+            }
+
+        // Calculate start and end IDs so there are no gaps
+        const types::particle_index start_id = llround(n_tracers * start_weight / global_function_integral) + 1;
+
+        return generate_particles_in_subdomain(cell_weights,global_function_integral,start_weight,n_tracers, 1.1 * start_id);
+      }
+
+      template <int dim>
+      std::vector<double>
+      ProbabilityDensityFunction<dim>::local_accumulated_cell_weights () const
+      {
+        //evaluate function at all cell midpoints, sort cells according to weight
+        const QMidpoint<dim> quadrature_formula;
+        const unsigned int n_quadrature_points = quadrature_formula.size();
+
+        FEValues<dim> fe_values (this->get_mapping(),
+                                 this->get_fe(),
+                                 quadrature_formula,
+                                 update_quadrature_points |
+                                 update_JxW_values);
+
+        std::vector<double> accumulated_cell_weights;
+
+        // compute the integral weight by quadrature
+        typename DoFHandler<dim>::active_cell_iterator
+        cell = this->get_dof_handler().begin_active(),
+        endc = this->get_dof_handler().end();
+        for (; cell!=endc; ++cell)
+          if (cell->is_locally_owned())
+            {
+              fe_values.reinit (cell);
+              const std::vector<Point<dim> > position = fe_values.get_quadrature_points();
+
+              // Then add the weight of the current cell. Weights are always
+              // interpreted positively, even if the function evaluates to a
+              // negative number.
+              double next_cell_weight = 0.0;
+              if (accumulated_cell_weights.size() > 0)
+                next_cell_weight = accumulated_cell_weights.back();
```

Comment:
Just pull the declaration of `next_cell_weight` out of the loop. You then don't need the `if` statement because you keep adding to the same variable and it automatically has the value of the previous cell as you get onto the next cell.